### ISSUE TITLE
use whatanime.ga to find unknown anime automatically

### ISSUE THREAD

**Issue by CuddleBear92** [ShokoAnime/ShokoServer#703]
https://whatanime.ga/ does search for anime after screenshots and has an api being built atm.
This would require the shoko server to grab a random screenshot or two and uploading that to the site or uploading to an image site and linking it (direct upload should be for the better).

if its made to use the site slowly and sparingly over a longer set of time. say less 3-5 screenshots per hour. made to slowly find the unknown anime.

Anime over a certain % on the site (say +85%) should be matched and shoko should pull the title of the series and episode number to be used locally.
Sad thing is that the site does not link to AniDB, MAL or most of the other anime sites.

this can help to dump more content on AniDB in the end as files can always be named wrongly or have Japanese titles which requires googling for most users.

what it would require:
- [ ] Random or % of length screenshots, a few each file (though one should be enough if its an interesting scene with characters) 

- [ ] Upload image to whatanime.ga (maybe though post or the api?)

- [ ] Get results and save them locally in the database. (if similarity is high enough)

- [ ] If match is found then remove it from the upload queue to not stress the server.

- [ ] If no match is found within the certain % of similarity across all the screenshots taken then take it out of queue and mark it for manual review.

- [ ] Automatically upload a very few files per hour to match made to slowly match over time to not stress the servers.

- [ ] Successfully marked files should be marked in the ui in the desktop client. With a simple click to take you to the correct avdump page.

issues with the site:
doesnt have direct links to metadata sites.
doesnt have much adult content in its database so matches with that type of content can be lacking (though there is some)

Links:
https://whatanime.ga/
https://github.com/soruly/whatanime.ga
api: https://soruly.github.io/whatanime.ga/ (broken page? or just locked with a paywall from patreon? unsure)
https://www.patreon.com/soruly

**Comment by da3dsoul** [comment]
This will still need to be per file, as we link file to episode. Due to the lack of info on release groups, it will need to be an extension of the Unrecognized Files Utility. It can be used to help manually link and AVDump files, though.

**Comment by CuddleBear92** [comment]
yeah it would need to be per file. don't see that much of an issue in the end if its made to do it over a longer amount of time slowly.
the biggest issue would be to stress their servers too much and they would block us.
but even at once file an hour or two will give you a series a day in the long run.
well if it gets a match on the first screenshot.

**Comment by CuddleBear92** [comment]
should it automatically add an empty series to the database when a match is made? would make sense to automatically do that too as it would cut the time to wait before its all in place when the user finally dumps and rechecks it.
and if it did that then it could add an custom tag to the series or a flag that we can filter.

**Comment by da3dsoul** [comment]
I'd say make it a setting

**Comment by CuddleBear92** [comment]
edited the whole OP to fit the talk that was on the discord server last night.
moving away from api usage and their site to doing some work on the user servers and sending it to the webcache to compare.

**Comment by Cazzar** [comment]
I wouldn't trust the MAL id especially if we wanted to pull data as the MAL
API tends to be unreliable

**Comment by da3dsoul** [comment]
You could use AniList with MAL IDs, though.

**Comment by ElementalCrisis** [comment]
Hey @soruly, WhatAnime has helped me find a few series so thanks for creating it. :) 

We're currently in the middle of doing some major changes to Shoko Server but I like the idea of using WhatAnime to help with our unrecognized process. Looking forward to that update to see what we can do with it. :)

**Comment by soruly** [comment]
In the long run I'd prefer to make whatanime.ga distributed, so anyone with skills would be able to setup an image database of their own.

**Comment by soruly** [comment]
With my new improvements made on searching, I think a decent quad core CPU would be able to handle any search in 5 seconds. And for hashing video, a 24-minute video takes ~30 seconds to hash on a 4GHz quad core machine). 

My plan is to open and publish my hashes (maybe like this https://data.whatanime.ga/100240/ ) and users just need to download and import these pre-hashed files into their own database for local search.

**Comment by misakitchi** [comment]
Good Idea! :)
"Distributed computing" like SETI

**Comment by soruly** [comment]
I've opensourced the distributed indexing system, you can take a look
https://github.com/soruly/sola

**Comment by Cazzar** [comment]
If there is a way we could port the video indexing into .net it could be something to have shoko, probably as an opt-in situation

**Comment by soruly** [comment]
whatanime.ga has moved to https://trace.moe
https://www.patreon.com/posts/moving-to-new-22212117

**Comment by mez0ru** [comment]
Any news on this? I have a lot of episodes that won't get unrecognized because of the hash system anidb uses. I would appreciate giving this a chance
